### PR TITLE
require teacher application when enrolling in certain workshops

### PR DESCRIPTION
**Pull request by davidsbailey**
Finishes https://codedotorg.atlassian.net/browse/ACQ-473. In addition to the stated requirements, also verifies that the application is for the current application year, but does not verify that the course name on the application matches the course associated with the workshop.

upon following the enrollment link, a teacher without an accepted application for the current year will see the following:

Screenshot 2023-03-21 at 9 27 06 AM

## Testing story

new unit tests covering new behavior. existing test coverage on enrollments controller covers existing behavior. adding UI tests for the enrollment page is covered by https://codedotorg.atlassian.net/browse/ACQ-481 .

### CODE REVIEW

File: dashboard/app/models/pd/workshop.rb [hunk 1]
```diff
@@ -128,6 +128,14 @@ def virtual_only_subjects_must_be_virtual
     end
   end
 
+  # Whether enrollment in this workshop requires an application
+  def require_application?
+    courses = [COURSE_CSP, COURSE_CSD, COURSE_CSA]
+    subjects = ACADEMIC_YEAR_SUBJECTS.concat([SUBJECT_SUMMER_WORKSHOP])
+    courses.include?(course) && subjects.include?(subject) &&
+      regional_partner && regional_partner.link_to_partner_application.blank?
```
Comment (megcrenshaw) [inline dashboard/app/models/pd/workshop.rb:136]: TIL about `.blank?`!

File: dashboard/app/views/pd/workshop_enrollment/missing_application.html.haml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+%h1
+  Teacher Application Is Required
+
+%p
+  You must have a completed application in order to enroll in this workshop. If
```
Comment (megcrenshaw) [inline dashboard/app/views/pd/workshop_enrollment/missing_application.html.haml]: Nit: It looks like the application must be not only complete, but also accepted and with an RP. Does the user "need" to know those details?

Comment (dmcavoy) [reply dashboard/app/views/pd/workshop_enrollment/missing_application.html.haml]: This is my bad. We should update the text here to say "You must have an accepted application...."

Comment (davidsbailey) [reply dashboard/app/views/pd/workshop_enrollment/missing_application.html.haml]: wording has been updated to mention "accepted". Meg, when you say "with an RP" are you referring to the workshop being associated with an RP? I don't think we need to clarify why the application is required, only what state the application needs to be in to satisfy the requirement.

**Review by megcrenshaw — APPROVED** [review]
Looks great!

**Pull request merged**